ADBDEV-4726-34 Check that elems is not null

dnskvlnk opened the pull request:
Check that elems is not null

construct_md_array passes elems to CopyArrayEls where it's dereferenced without
checking that it's not null. We can't guarantee that elems is always valid.
This patch checks that it won't happen

Commit:
Check that vlues is not null if nitems > 0

Commit:
Check that values is not null if nitems > 0

CopyArrayEls dereferences values without checking that it's not null.
This patch adds an assert if nitems != 0

RekGRpth commented on src/backend/utils/adt/arrayfuncs.c:
```diff
@@ -945,6 +945,8 @@ CopyArrayEls(ArrayType *array,
 	int			bitmask = 1;
 	int			i;
 
+	Assert(values || nitems == 0);
```
I think there is no reason to call CopyArrayEls with NULL as values and 0 as nitems

dnskvlnk replied:
Removed == 0 condition and switched to Insist

Commit:
Check that values is not null

CopyArrayEls dereferences values without checking that it's not null.
We can't guarantee that values is always valid. This patch adds Insist

RekGRpth commented on src/backend/utils/adt/arrayfuncs.c:
```diff
@@ -945,6 +945,8 @@ CopyArrayEls(ArrayType *array,
 	int			bitmask = 1;
 	int			i;
 
+	Insist(values);
```
I think you should instead rewrite the conditions here somehow: https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/utils/adt/arrayfuncs.c#L3063-L3073

dnskvlnk replied:
Why? `values` must not be null in `CopyArrayEls`, so I think it's better to add this condition there

RekGRpth replied:
1) elems = NULL
2) fixedwidthtype = false
3) call CopyArrayEls with NULL as values

may be condition should be
```c
if (elems==NULL || fixedwidthtype)
{
/* do nothing */
}
```
?

dnskvlnk replied:
It seems like here we allocate a new array to fill it with CopyArrayEls, so I think your proposal is correct
https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/utils/adt/arrayfuncs.c#L3021-L3042

dnskvlnk replied:
Done

dnskvlnk replied:
I think it should be like 
```
if (elems==NULL && fixedwidthtype)
{
/* do nothing */
} else if (elems != NULL) {
    CopyArrayEls
```
However, the program shouldn't run any further if elems == NULL and !fixedwidthtype so it may be also `Insist`

RekGRpth replied:
failed tests: uao_dml/uao_dml_row uao_dml/uao_dml_column

dnskvlnk replied:
I think we should just use Insist in else block

RekGRpth replied:
> if (elems==NULL && fixedwidthtype)
> {
> /* do nothing */
> } else if (elems != NULL) {
>     CopyArrayEls

this solution is working!

dnskvlnk replied:
But what would happen if elems == NULL and !fixedwidthtype? Doesn't seem like it can be tolerated

Commit:
Check that elems is not null

construct_md_array passes elems to CopyArrayEls where it's dereferenced without
checking that it's not null. We can't guarantee that elems is always valid.
This patch checks that it won't happen

bandetto commented on src/backend/utils/adt/arrayfuncs.c:
```diff
@@ -3060,7 +3060,7 @@ construct_md_array(Datum *elems,
 	memcpy(ARR_LBOUND(result), lbs, ndims * sizeof(int));
 
-	if (elems==NULL && fixedwidthtype) 
+	if (elems==NULL || fixedwidthtype) 
```
I suggest a change. It seems weird that we don't copy elements if they are of fixed width.
```suggestion
	if (elems==NULL)
```
There is some info about this in the comment atop of this function but I found it a little confusing. Also, `fixedwidthtype` is Greenplum-specific code.

RekGRpth replied:
> I suggest a change. It seems weird that we don't copy elements if they are of fixed width.
> 
> There is some info about this in the comment atop of this function but I found it a little confusing. Also, `fixedwidthtype` is Greenplum-specific code.

Also working idea
